README
PHP-gettext 1.0

Copyright 2003, 2006, 2009 -- Danilo "angry with PHP[1]" Segan
Licensed under GPLv2 (or any later version, see COPYING)

[1] PHP is actually cyrillic, and translates roughly to
    "works-doesn't-work" (UTF-8: Ради-Не-Ради)


Introduction

    How many times did you look for a good translation tool, and
    found out that gettext is best for the job? Many times.

    How many times did you try to use gettext in PHP, but failed
    miserably, because either your hosting provider didn't support
    it, or the server didn't have adequate locale? Many times.

    Well, this is a solution to your needs. It allows using gettext
    tools for managing translations, yet it doesn't require gettext
    library at all. It parses generated MO files directly, and thus
    might be a bit slower than the (maybe provided) gettext library.

    PHP-gettext is a simple reader for GNU gettext MO files. Those
    are binary containers for translations, produced by GNU msgfmt.

Why?

    I got used to having gettext work even without gettext
    library. It's there in my favourite language Python, so I was
    surprised that I couldn't find it in PHP. I even Googled for it,
    but to no avail.

    So, I said, what the heck, I'm going to write it for this
    disguisting language of PHP, because I'm often constrained to it.

Features

  o Support for simple translations
    Just define a simple alias for translate() function (suggested
    use of _() or gettext(); see provided example).

  o Support for ngettext calls (plural forms, see a note under bugs)
    You may also use plural forms. Translations in MO files need to
    provide this, and they must also provide "plural-forms" header.
    Please see 'info gettext' for more details.

  o Support for reading straight files, or strings (!!!)
    Since I can imagine many different backends for reading in the MO
    file data, I used imaginary abstract class StreamReader to do all
    the input (check streams.php). For your convenience, I've already
    provided two classes for reading files: FileReader and
    StringReader (CachedFileReader is a combination of the two: it
    loads entire file contents into a string, and then works on that).
    See example below for usage. You can for instance use StringReader
    when you read in data from a database, or you can create your own
    derivative of StreamReader for anything you like.


Bugs

    Report them on https://bugs.launchpad.net/php-gettext

Usage

    Put files streams.php and gettext.php somewhere you can load them
    from, and require 'em in where you want to use them.

    Then, create one 'stream reader' (a class that provides functions
    like read(), seekto(), currentpos() and length()) which will
    provide data for the 'gettext_reader', with eg.
      $streamer = new FileStream('data.mo');

    Then, use that as a parameter to gettext_reader constructor:
      $wohoo = new gettext_reader($streamer);

    If you want to disable pre-loading of entire message catalog in
    memory (if, for example, you have a multi-thousand message catalog
    which you'll use only occasionally), use "false" for second
    parameter to gettext_reader constructor:
      $wohoo = new gettext_reader($streamer, false);

    From now on, you have all the benefits of gettext data at your
    disposal, so may run:
      print $wohoo->translate("This is a test");
      print $wohoo->ngettext("%d bird", "%d birds", $birds);

    You might need to pass parameter "-k" to xgettext to make it
    extract all the strings. In above example, try with
      xgettext -ktranslate -kngettext:1,2 file.php
    what should create messages.po which contains two messages for
    translation.

    I suggest creating simple aliases for these functions (see
    example/pigs.php for how do I do it, which means it's probably a
    bad way).


Usage with gettext.inc (standard gettext interfaces emulation)

    Check example in examples/pig_dropin.php, basically you include
    gettext.inc and use all the standard gettext interfaces as
    documented on:

       http://www.php.net/gettext

    The only catch is that you can check return value of setlocale()
    to see if your locale is system supported or not.


Example

    See in examples/ subdirectory. There are a couple of files.
    pigs.php is an example, serbian.po is a translation to Serbian
    language, and serbian.mo is generated with
       msgfmt -o serbian.mo serbian.po
    There is also simple "update" script that can be used to generate
    POT file and to update the translation using msgmerge.

TODO:

  o Improve speed to be even more comparable to the native gettext
    implementation.

  o Try to use hash tables in MO files: with pre-loading, would it
    be useful at all?

Never-asked-questions:

  o Why did you mark this as version 1.0 when this is the first code
    release?

    Well, it's quite simple. I consider that the first released thing
    should be labeled "version 1" (first, right?). Zero is there to
    indicate that there's zero improvement and/or change compared to
    "version 1".

    I plan to use version numbers 1.0.* for small bugfixes, and to
    release 1.1 as "first stable release of version 1".

    This may trick someone that this is actually useful software, but
    as with any other free software, I take NO RESPONSIBILITY for
    creating such a masterpiece that will smoke crack, trash your
    hard disk, and make lasers in your CD device dance to the tune of
    Mozart's 40th Symphony (there is one like that, right?).

  o Can I...?

    Yes, you can. This is free software (as in freedom, free speech),
    and you might do whatever you wish with it, provided you do not
    limit freedom of others (GPL).

    I'm considering licensing this under LGPL, but I *do* want
    *every* PHP-gettext user to contribute and respect ideas of free
    software, so don't count on it happening anytime soon.

    I'm sorry that I'm taking away your freedom of taking others'
    freedom away, but I believe that's neglible as compared to what
    freedoms you could take away. ;-)

    Uhm, whatever.


ChangeLog
2006-02-28  Danilo Šegan  <danilo@gnome.org>

	Fixes bug #15923.
	
	* gettext.php (gettext_reader): make magic check work on 64-bit
	platforms as well (by Steffen Pingel).

2006-02-20  Danilo Šegan  <danilo@gnome.org>

	* gettext.inc (_bindtextdomain): Use php_uname to detect Windows.

2006-02-07  Danilo Šegan  <danilo@gnome.org>

	* examples/pigs_dropin.php: comment-out bind_textdomain_codeset

	* gettext.inc (T_bind_textdomain_codeset): bind_textdomain_codeset
	is available only in PHP 4.2.0+ (thanks to Jens A. Tkotz).

	* Makefile: Include gettext.inc in DIST_FILES, VERSION up to
	1.0.7.

2006-02-03  Danilo Šegan  <danilo@gnome.org>

	Added setlocale() emulation as well.
	
	* examples/pigs_dropin.php: Use T_setlocale() and locale_emulation().
	* examples/pigs_fallback.php: Use T_setlocale() and locale_emulation().

	* gettext.inc: 	Added globals $EMULATEGETTEXT and $CURRENTLOCALE.
	(locale_emulation): Whether emulation is active.
	(_check_locale): Rewrite.
	(_setlocale): Added emulated setlocale function.
	(T_setlocale): Wrapper around _setlocale.
	(_get_reader): Use variables and _setlocale.

2006-02-02  Danilo Šegan  <danilo@gnome.org>

	Fix bug #12192.

	* examples/locale/sr_CS/LC_MESSAGES/messages.po: Correct grammar.
	* examples/locale/sr_CS/LC_MESSAGES/messages.mo: Rebuild.

2006-02-02  Danilo Šegan  <danilo@gnome.org>

	Fix bug #15419.
	
	* streams.php: Support for PHP 5.1.1 fread() which reads most 8kb.
	(Fix by Piotr Szotkowski <shot@hot.pl>)

2006-02-02  Danilo Šegan  <danilo@gnome.org>

	Merge Steven Armstrong's changes, supporting standard gettext
	interfaces: 

	* examples/*: Restructured examples.
	* gettext.inc: Added.
	* AUTHORS: Added Steven.
	* Makefile (VERSION): Up to 1.0.6.
	
2006-01-28  Nico Kaiser  <nico@siriux.net>

	* gettext.php (select_string): Fix "true" <-> 1 difference of PHP

2005-07-29  Danilo Šegan  <danilo@gnome.org>

	* Makefile (VERSION): Up to 1.0.5.

2005-07-29  Danilo Šegan  <danilo@gnome.org>

	Fixes bug #13850.
	
	* gettext.php (gettext_reader): check $Reader->error as well.

2005-07-29  Danilo Šegan  <danilo@gnome.org>

	* Makefile (VERSION): Up to 1.0.4.

2005-07-29  Danilo Šegan  <danilo@gnome.org>

	Fixes bug #13771.

	* gettext.php (gettext_reader->get_plural_forms): Plural forms
	header extraction regex change.  Reported by Edgar Gonzales.
	
2005-02-28  Danilo Šegan  <dsegan@gmx.net>

	* AUTHORS: Added Nico to the list.

	* Makefile (VERSION): Up to 1.0.3.

	* README: Updated.

2005-02-28  Danilo Šegan  <dsegan@gmx.net>

	* gettext.php: Added pre-loading, code documentation, and many
	code clean-ups by Nico Kaiser <nico@siriux.net>.

2005-02-28  Danilo Šegan  <dsegan@gmx.net>

	* streams.php (FileReader.read): Handle read($bytes = 0).

	* examples/pigs.php: Prefix gettext function names with T or T_.

	* examples/update: Use the same keywords T_ and T_ngettext.

	* streams.php: Added CachedFileReader.

2003-11-11  Danilo Šegan  <dsegan@gmx.net>

	* gettext.php: Added hashing to find_string.

2003-11-01  Danilo Šegan  <dsegan@gmx.net>

	* Makefile (DIST_FILES): Replaced LICENSE with COPYING.
	(VERSION): Up to 1.0.2.

	* AUTHORS: Minor edits.

	* README: Minor edits.

	* COPYING: Removed LICENSE, added this file.

	* gettext.php: Added copyright notice and disclaimer.
	* streams.php: Same.
	* examples/pigs.php: Same.

2003-10-23  Danilo Šegan  <dsegan@gmx.net>

	* Makefile: Upped version to 1.0.1.

	* gettext.php (gettext_reader): Remove a call to set_total_plurals.
	(set_total_plurals): Removed unused function for some better days.

2003-10-23  Danilo Šegan  <dsegan@gmx.net>

	* Makefile: Added, version 1.0.0.

	* examples/*: Added an example of usage.

	* README: Described all the crap.
	
2003-10-22  Danilo Šegan  <dsegan@gmx.net>

	* gettext.php: Plural forms implemented too.

	* streams.php: Added FileReader for direct access to files (no
	need to keep file in memory).

	* gettext.php: It works, except for plural forms.

	* streams.php: Created abstract class StreamReader.
	Added StringReader class.

	* gettext.php: Started writing gettext_reader.



gettext.php
<?php
/*
   Copyright (c) 2003 Danilo Segan <danilo@kvota.net>.
   Copyright (c) 2005 Nico Kaiser <nico@siriux.net>
   
   This file is part of PHP-gettext.

   PHP-gettext is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   PHP-gettext is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with PHP-gettext; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*/
 
/**
 * Provides a simple gettext replacement that works independently from
 * the system's gettext abilities.
 * It can read MO files and use them for translating strings.
 * The files are passed to gettext_reader as a Stream (see streams.php)
 * 
 * This version has the ability to cache all strings and translations to
 * speed up the string lookup.
 * While the cache is enabled by default, it can be switched off with the
 * second parameter in the constructor (e.g. whenusing very large MO files
 * that you don't want to keep in memory)
 */
class gettext_reader {
  //public:
   var $error = 0; // public variable that holds error code (0 if no error)
   
   //private:
  var $BYTEORDER = 0;        // 0: low endian, 1: big endian
  var $STREAM = NULL;
  var $short_circuit = false;
  var $enable_cache = false;
  var $originals = NULL;      // offset of original table
  var $translations = NULL;    // offset of translation table
  var $pluralheader = NULL;    // cache header field for plural forms
  var $total = 0;          // total string count
  var $table_originals = NULL;  // table for original strings (offsets)
  var $table_translations = NULL;  // table for translated strings (offsets)
  var $cache_translations = NULL;  // original -> translation mapping


  /* Methods */
  
    
  /**
   * Reads a 32bit Integer from the Stream
   * 
   * @access private
   * @return Integer from the Stream
   */
  function readint() {
      if ($this->BYTEORDER == 0) {
        // low endian
        return array_shift(unpack('V', $this->STREAM->read(4)));
      } else {
        // big endian
        return array_shift(unpack('N', $this->STREAM->read(4)));
      }
    }

  /**
   * Reads an array of Integers from the Stream
   * 
   * @param int count How many elements should be read
   * @return Array of Integers
   */
  function readintarray($count) {
    if ($this->BYTEORDER == 0) {
        // low endian
        return unpack('V'.$count, $this->STREAM->read(4 * $count));
      } else {
        // big endian
        return unpack('N'.$count, $this->STREAM->read(4 * $count));
      }
  }
  
  /**
   * Constructor
   * 
   * @param object Reader the StreamReader object
   * @param boolean enable_cache Enable or disable caching of strings (default on)
   */
  function gettext_reader($Reader, $enable_cache = true) {
    // If there isn't a StreamReader, turn on short circuit mode.
    if (! $Reader || isset($Reader->error) ) {
      $this->short_circuit = true;
      return;
    }
    
    // Caching can be turned off
    $this->enable_cache = $enable_cache;

    // $MAGIC1 = (int)0x950412de; //bug in PHP 5
    $MAGIC1 = (int) - 1794895138;
    // $MAGIC2 = (int)0xde120495; //bug
    $MAGIC2 = (int) - 569244523;

    $this->STREAM = $Reader;
    $magic = $this->readint();
    if ($magic == ($MAGIC1 & 0xFFFFFFFF)) {
      $this->BYTEORDER = 0;
    } elseif ($magic == ($MAGIC2 & 0xFFFFFFFF)) {
      $this->BYTEORDER = 1;
    } else {
      $this->error = 1; // not MO file
      return false;
    }
    
    // FIXME: Do we care about revision? We should.
    $revision = $this->readint();
    
    $this->total = $this->readint();
    $this->originals = $this->readint();
    $this->translations = $this->readint();
  }
  
  /**
   * Loads the translation tables from the MO file into the cache
   * If caching is enabled, also loads all strings into a cache
   * to speed up translation lookups
   * 
   * @access private
   */
  function load_tables() {
    if (is_array($this->cache_translations) &&
      is_array($this->table_originals) &&
      is_array($this->table_translations))
      return;
    
    /* get original and translations tables */
    $this->STREAM->seekto($this->originals);
    $this->table_originals = $this->readintarray($this->total * 2);
    $this->STREAM->seekto($this->translations);
    $this->table_translations = $this->readintarray($this->total * 2);
    
    if ($this->enable_cache) {
      $this->cache_translations = array ();
      /* read all strings in the cache */
      for ($i = 0; $i < $this->total; $i++) {
        $this->STREAM->seekto($this->table_originals[$i * 2 + 2]);
        $original = $this->STREAM->read($this->table_originals[$i * 2 + 1]);
        $this->STREAM->seekto($this->table_translations[$i * 2 + 2]);
        $translation = $this->STREAM->read($this->table_translations[$i * 2 + 1]);
        $this->cache_translations[$original] = $translation;
      }
    }
  }
  
  /**
   * Returns a string from the "originals" table
   * 
   * @access private
   * @param int num Offset number of original string
   * @return string Requested string if found, otherwise ''
   */
  function get_original_string($num) {
    $length = $this->table_originals[$num * 2 + 1];
    $offset = $this->table_originals[$num * 2 + 2];
    if (! $length)
      return '';
    $this->STREAM->seekto($offset);
    $data = $this->STREAM->read($length);
    return (string)$data;
  }
  
  /**
   * Returns a string from the "translations" table
   * 
   * @access private
   * @param int num Offset number of original string
   * @return string Requested string if found, otherwise ''
   */
  function get_translation_string($num) {
    $length = $this->table_translations[$num * 2 + 1];
    $offset = $this->table_translations[$num * 2 + 2];
    if (! $length)
      return '';
    $this->STREAM->seekto($offset);
    $data = $this->STREAM->read($length);
    return (string)$data;
  }
  
  /**
   * Binary search for string
   * 
   * @access private
   * @param string string
   * @param int start (internally used in recursive function)
   * @param int end (internally used in recursive function)
   * @return int string number (offset in originals table)
   */
  function find_string($string, $start = -1, $end = -1) {
    if (($start == -1) or ($end == -1)) {
      // find_string is called with only one parameter, set start end end
      $start = 0;
      $end = $this->total;
    }
    if (abs($start - $end) <= 1) {
      // We're done, now we either found the string, or it doesn't exist
      $txt = $this->get_original_string($start);
      if ($string == $txt)
        return $start;
      else
        return -1;
    } else if ($start > $end) {
      // start > end -> turn around and start over
      return $this->find_string($string, $end, $start);
    } else {
      // Divide table in two parts
      $half = (int)(($start + $end) / 2);
      $cmp = strcmp($string, $this->get_original_string($half));
      if ($cmp == 0)
        // string is exactly in the middle => return it
        return $half;
      else if ($cmp < 0)
        // The string is in the upper half
        return $this->find_string($string, $start, $half);
      else
        // The string is in the lower half
        return $this->find_string($string, $half, $end);
    }
  }
  
  /**
   * Translates a string
   * 
   * @access public
   * @param string string to be translated
   * @return string translated string (or original, if not found)
   */
  function translate($string) {
    if ($this->short_circuit)
      return $string;
    $this->load_tables();     
    
    if ($this->enable_cache) {
      // Caching enabled, get translated string from cache
      if (array_key_exists($string, $this->cache_translations))
        return $this->cache_translations[$string];
      else
        return $string;
    } else {
      // Caching not enabled, try to find string
      $num = $this->find_string($string);
      if ($num == -1)
        return $string;
      else
        return $this->get_translation_string($num);
    }
  }

  /**
   * Get possible plural forms from MO header
   * 
   * @access private
   * @return string plural form header
   */
  function get_plural_forms() {
    // lets assume message number 0 is header  
    // this is true, right?
    $this->load_tables();
    
    // cache header field for plural forms
    if (! is_string($this->pluralheader)) {
      if ($this->enable_cache) {
        $header = $this->cache_translations[""];
      } else {
        $header = $this->get_translation_string(0);
      }
      if (eregi("plural-forms: ([^\n]*)\n", $header, $regs))
        $expr = $regs[1];
      else
        $expr = "nplurals=2; plural=n == 1 ? 0 : 1;";
      $this->pluralheader = $expr;
    }
    return $this->pluralheader;
  }

  /**
   * Detects which plural form to take
   * 
   * @access private
   * @param n count
   * @return int array index of the right plural form
   */
  function select_string($n) {
    $string = $this->get_plural_forms();
    $string = str_replace('nplurals',"\$total",$string);
    $string = str_replace("n",$n,$string);
    $string = str_replace('plural',"\$plural",$string);
    
    $total = 0;
    $plural = 0;

    eval("$string");
    if ($plural >= $total) $plural = $total - 1;
    return $plural;
  }

  /**
   * Plural version of gettext
   * 
   * @access public
   * @param string single
   * @param string plural
   * @param string number
   * @return translated plural form
   */
  function ngettext($single, $plural, $number) {
    if ($this->short_circuit) {
      if ($number != 1)
        return $plural;
      else
        return $single;
    }

    // find out the appropriate form
    $select = $this->select_string($number); 
    
    // this should contains all strings separated by NULLs
    $key = $single.chr(0).$plural;
    
    
    if ($this->enable_cache) {
      if (! array_key_exists($key, $this->cache_translations)) {
        return ($number != 1) ? $plural : $single;
      } else {
        $result = $this->cache_translations[$key];
        $list = explode(chr(0), $result);
        return $list[$select];
      }
    } else {
      $num = $this->find_string($key);
      if ($num == -1) {
        return ($number != 1) ? $plural : $single;
      } else {
        $result = $this->get_translation_string($num);
        $list = explode(chr(0), $result);
        return $list[$select];
      }
    }
  }

}

?>


examples/pigs_dropin.php
<?php
/*
   Copyright (c) 2003,2004,2005 Danilo Segan <danilo@kvota.net>.
   Copyright (c) 2005,2006 Steven Armstrong <sa@c-area.ch>

   This file is part of PHP-gettext.

   PHP-gettext is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   PHP-gettext is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with PHP-gettext; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*/

// define constants
define(PROJECT_DIR, realpath('./'));
define(LOCALE_DIR, PROJECT_DIR .'/locale');
define(DEFAULT_LOCALE, 'en_US');

require_once('../gettext.inc');

$supported_locales = array('en_US', 'sr_CS', 'de_CH');
$encoding = 'UTF-8';

$locale = (isset($_GET['lang']))? $_GET['lang'] : DEFAULT_LOCALE;

// gettext setup
putenv("LANG=$locale");
T_setlocale(LC_MESSAGES, $locale);
// Set the text domain as 'messages'
$domain = 'messages';
bindtextdomain($domain, LOCALE_DIR);
bind_textdomain_codeset($domain, $encoding);
textdomain($domain);

header("Content-type: text/html; charset=$encoding");
?>
<html>
<head>
<title>PHP-gettext dropin example</title>
</head>
<body>
<h1>PHP-gettext as a dropin replacement</h1>
<p>Example showing how to use PHP-gettext as a dropin replacement for the native gettext library.</p>
<?php
print "<p>";
foreach($supported_locales as $l) {
	print "[<a href=\"?lang=$l\">$l</a>] ";
}
print "</p>\n";

if (!locale_emulation()) {
	print "<p>locale '$locale' is supported by your system, using native gettext implementation.</p>\n";
}
else {
	print "<p>locale '$locale' is _not_ supported on your system, using the default locale '". DEFAULT_LOCALE ."'.</p>\n";
}
?>

<hr />

<?php
// using PHP-gettext
print "<pre>";
print _("This is how the story goes.\n\n");
for ($number=6; $number>=0; $number--) {
  print sprintf(T_ngettext("%d pig went to the market\n", 
			  "%d pigs went to the market\n", $number), 
		 $number );
}
print "</pre>\n";
?>

<hr />
<p>&laquo; <a href="./">back</a></p>
</body>
</html>


examples/pigs_fallback.php
<?php
/*
   Copyright (c) 2003,2004,2005 Danilo Segan <danilo@kvota.net>.
   Copyright (c) 2005,2006 Steven Armstrong <sa@c-area.ch>

   This file is part of PHP-gettext.

   PHP-gettext is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   PHP-gettext is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with PHP-gettext; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*/

// define constants
define(PROJECT_DIR, realpath('./'));
define(LOCALE_DIR, PROJECT_DIR .'/locale');
define(DEFAULT_LOCALE, 'en_US');

require_once('../gettext.inc');

$supported_locales = array('en_US', 'sr_CS', 'de_CH');
$encoding = 'UTF-8';

$locale = (isset($_GET['lang']))? $_GET['lang'] : DEFAULT_LOCALE;

// gettext setup
putenv("LANG=$locale");
setlocale(LC_MESSAGES, $locale);
// Set the text domain as 'messages'
$domain = 'messages';
T_bindtextdomain($domain, LOCALE_DIR);
T_bind_textdomain_codeset($domain, $encoding);
T_textdomain($domain);

header("Content-type: text/html; charset=$encoding");
?>
<html>
<head>
<title>PHP-gettext fallback example</title>
</head>
<body>
<h1>PHP-gettext as a fallback solution</h1>
<p>Example showing how to use PHP-gettext as a fallback solution if the native gettext library is not available or the system does not support the requested locale.</p>

<?php
print "<p>";
foreach($supported_locales as $l) {
	print "[<a href=\"?lang=$l\">$l</a>] ";
}
print "</p>\n";

if (setlocale(LC_MESSAGES, '')) {
	print "<p>locale '$locale' is supported by your system, using native gettext implementation.</p>\n";
}
else {
	print "<p>locale '$locale' is <strong>not</strong> supported on your system, using custom gettext implementation.</p>\n";
}
?>

<hr />

<?php
// using PHP-gettext
print "<pre>";
print T_("This is how the story goes.\n\n");
for ($number=6; $number>=0; $number--) {
  print sprintf( T_ngettext("%d pig went to the market\n", 
			  "%d pigs went to the market\n", $number), 
		 $number );
}
print "</pre>\n";
?>

<hr />
<p>&laquo; <a href="./">back</a></p>
</body>
</html>
